charmstore/fakeclient.go

```diff
@@ -218,7 +385,117 @@ func (r Repository) Get(id *charm.URL) (charm.Charm, error) {
 func (r Repository) GetBundle(id *charm.URL) (charm.Bundle, error) {
 	bundleData := r.bundles[r.channel][*id]
 	if bundleData == nil {
-		return bundleData, errors.NotFoundf(id.String())
+		return nil, errors.NotFoundf(id.String())
 	}
 	return bundleData, nil
 }
+
+// ListResources returns Resource metadata for resources that have been
+// uploaded to the repository for id. To upload a resource, use UploadResource.
+// Although id is type *charm.URL, resources are not restricted to charms. That
+// type is also used for other entities in the charmstore, such as bundles.
+//
+// Returns an error that satisfies errors.IsNotFound when no resources
+// are present for id.
+func (r Repository) ListResources(id *charm.URL) ([]params.Resource, error) {
+	resources := r.resources[r.channel][*id]
+	if len(resources) == 0 {
+		return resources, errors.NotFoundf("no resources for %v", id)
+	}
+	return resources, nil
+}
+
+func (r Repository) UploadCharm(id *charm.URL, charmData charm.Charm) (*charm.URL, error) {
+	if len(r.charms[r.channel]) == 0 {
+		r.charms[r.channel] = make(map[charm.URL]charm.Charm)
+	}
+	withRevision := r.addRevision(id)
+	r.charms[r.channel][*withRevision] = charmData
+	return withRevision, nil
+}
+
+func (r Repository) UploadCharmWithRevision(id *charm.URL, charmData charm.Charm, promulgatedRevision int) error {
+	if len(r.revisions[r.channel]) == 0 {
+		r.revisions[r.channel] = make(map[charm.URL]int)
+	}
+	r.revisions[r.channel][*id] = promulgatedRevision
+	_, err := r.UploadCharm(id, charmData)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (r Repository) UploadBundle(id *charm.URL, bundleData charm.Bundle) (*charm.URL, error) {
+	r.bundles[r.channel][*id] = bundleData
+	return id, nil
+}
+
+func (r Repository) UploadBundleWithRevision(id *charm.URL, bundleData charm.Bundle, promulgatedRevision int) error {
+	_, err := r.UploadBundle(id, bundleData)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	r.revisions[r.channel][*id] = promulgatedRevision
+	return nil
+}
+
+// UploadResource "uploads" data from file and stores it at path
+func (r Repository) UploadResource(id *charm.URL, name, path string, file io.ReaderAt, size int64, progress csclient.Progress) (revision int, err error) {
+	if len(r.resources[r.channel]) == 0 {
+		r.resources[r.channel] = make(map[charm.URL][]params.Resource)
+	}
+	resources, err := r.ListResources(id)
+	if err != nil {
+		return -1, errors.Trace(err)
+	}
+	revision = len(resources)
+	if len(resources) == 0 {
+		r.resources[r.channel][*id] = make([]params.Resource, 1)
```

Comment:
no need to preallocate a slice, just append below
(change anywhere else as well)

Reply:
Done